datafusion/optimizer/src/eliminate_duplicated_expr.rs

```diff
@@ -79,14 +51,15 @@ impl OptimizerRule for EliminateDuplicatedExpr {
         match plan {
             LogicalPlan::Sort(sort) => {
                 let len = sort.expr.len();
-                let unique_exprs: Vec<_> = sort
-                    .expr
-                    .into_iter()
-                    .map(|e| SortExprWrapper { expr: e })
-                    .collect::<IndexSet<_>>()
-                    .into_iter()
-                    .map(|wrapper| wrapper.expr)
-                    .collect();
+                let mut first_sort_by_expr: IndexMap<Expr, SortExpr> =
+                    IndexMap::default();
+                for s in &sort.expr {
```

Comment:
I think we could avoid clone here, take `expr`

Reply:
i wasn't able to fully eliminate clone here.
see `reduce cloning in EliminateDuplicatedExpr` commit -- this is how far i got.
can you please advise on how to improve it?

i though also about using `IndexMap<&Expr, usize>` to gather indexes of surviving sort expressions, and then using that to filter the `Vec<Sort>`. This would require hashing twice (first expressions, then surviving positions) and still would involve copying for the output vec. so not sure this would help overall, without measuring

Reply:
I take a look again, I think this is what we need

```rust
LogicalPlan::Sort(sort) => {
    let len = sort.expr.len();
    let unique_exprs: IndexSet<_> = sort.expr.into_iter().map(|e| SortExprWrapper(e)).collect();
    let unique_exprs = unique_exprs.into_iter().map(|e|e.0).collect::<Vec<_>>();

    let transformed = if len != unique_exprs.len() {
        Transformed::yes
    } else {
        Transformed::no
    };

    Ok(transformed(LogicalPlan::Sort(Sort {
        expr: unique_exprs,
        input: sort.input,
        fetch: sort.fetch,
    })))
}

#[derive(Eq)]
struct SortExprWrapper(SortExpr);

impl PartialEq for SortExprWrapper {
    fn eq(&self, other: &Self) -> bool {
        self.0.expr == other.0.expr
    }
}

impl Hash for SortExprWrapper {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.expr.hash(state)
    }
}
```

Reply:
thanks @jayzhan211! i see my mistake. 
there was a SortExprWrapper struct, but i did remove it
i restored SortExprWrapper code and adapted it. thus the code is different than in your comment, but the purpose was to make the diff smaller.

see [restore SortExprWrapper](https://github.com/apache/datafusion/pull/12177/commits/9931e1c718139a84806c4e83530d6f2a27cb78be) commit
or the full diff for that file https://github.com/apache/datafusion/pull/12177/files#diff-4740b5caba0e0f9addf958e3987a0f7745ab6f45e13dc213fec0e1304d057f0cL41
as the result there are only minimal changes in `SortExprWrapper` and no changes within `impl OptimizerRule for EliminateDuplicatedExpr`

Reply:
Don't optimize for the smaller diff.
I think self.0.expr better than self.expr.expr